lambda/translmod.ml

```diff
@@ -794,17 +785,27 @@ let required_globals ~flambda body =
 
 (* Compile an implementation *)
 
+let module_block_size component_names coercion =
+  match coercion with
+  | Tcoerce_none -> List.length component_names
+  | Tcoerce_structure (l, _) -> List.length l
+  | Tcoerce_functor _
+  | Tcoerce_primitive _
+  | Tcoerce_alias _ -> assert false
+
 let transl_implementation_flambda module_name (str, cc) =
   reset_labels ();
   primitive_declarations := [];
   Translprim.clear_used_primitives ();
   let module_id = Ident.create_persistent module_name in
   let scopes = enter_module_definition ~scopes:empty_scopes module_id in
-  let body, size =
+  let body =
     Translobj.transl_label_init
       (fun () -> transl_struct ~scopes Loc_unknown [] cc
                    (global_path module_id) str)
   in
+  let size =
+    module_block_size (bound_value_identifiers str.str_type) cc in
```

Comment:
Note: computing the set of bound identifiers again is not completely free, and computing it from the type is quite different from the previous approach of computing it from the implementation, but I think that the size from the type is more obviously correct and removing the tuple allocations likely saves up roughly as much as the cost of recomputing the set of identifiers.

Reply:
Indeed. If performance was critical, one could have a specialized version of `module_block_size` that would avoid computing the intermediate list in `bound_value_identifiers`, but it is probably not worth it at this time.